whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java

```diff
@@ -0,0 +1,99 @@
+
+
+@Service
+public class SnapshotFileGenerator {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SnapshotFileGenerator.class);
+
+    private final NrtmVersionInfoRepository nrtmVersionInfoRepository;
+    private final SnapshotFileSerializer snapshotFileSerializer;
+    private final SnapshotFileRepository snapshotFileRepository;
+    private final SnapshotObjectSynchronizer snapshotObjectSynchronizer;
+    private final NrtmFileStore nrtmFileStore;
+
+    public SnapshotFileGenerator(
+        final NrtmVersionInfoRepository nrtmVersionInfoRepository,
+        final SnapshotFileSerializer snapshotFileSerializer,
+        final SnapshotFileRepository snapshotFileRepository,
+        final SnapshotObjectSynchronizer snapshotObjectSynchronizer,
+        final NrtmFileStore nrtmFileStore
+    ) {
+        this.nrtmVersionInfoRepository = nrtmVersionInfoRepository;
+        this.snapshotFileSerializer = snapshotFileSerializer;
+        this.snapshotFileRepository = snapshotFileRepository;
+        this.snapshotObjectSynchronizer = snapshotObjectSynchronizer;
+        this.nrtmFileStore = nrtmFileStore;
+    }
+
+    public Optional<PublishableSnapshotFile> createSnapshot(final NrtmSource source) {
+        final String method = "createSnapshot";
+        // Get last version from database.
+        final Optional<NrtmVersionInfo> lastVersion = nrtmVersionInfoRepository.findLastVersion(source);
+        NrtmVersionInfo version;
+        LOGGER.info("{} lastVersion.isEmpty() {}", method, lastVersion.isEmpty());
+        if (lastVersion.isEmpty()) {
+            version = snapshotObjectSynchronizer.initializeSnapshotObjects(source);
+        } else {
+            version = lastVersion.get();
+            if (version.getType() == NrtmDocumentType.DELTA) {
+                LOGGER.info("Not generating snapshot file yet");
+                return Optional.empty();
+            } else {
+                LOGGER.info("{} Not generating snapshot file since no deltas have been published since v{} with serialID {}",
+                    method, version.getVersion(), version.getLastSerialId());
+                return Optional.empty();
+            }
+        }
+        LOGGER.info("{} now at version: {}", method, version);
+        if (version.getVersion() > 1) {
+            LOGGER.debug("not syncing deltas to snapshot");
+            return Optional.empty();
+        }
+        final PublishableSnapshotFile snapshotFile = new PublishableSnapshotFile(version);
+        final String fileName = NrtmFileUtil.newFileName(snapshotFile);
+        try (final OutputStream out = nrtmFileStore.getFileOutputStream(snapshotFile.getSessionID(), fileName)) {
+            final Stopwatch stopwatch = Stopwatch.createStarted();
+            snapshotFileSerializer.writeSnapshotAsJson(snapshotFile, out);
+            final String sha256hex = DigestUtils.sha256Hex(nrtmFileStore.getFileInputStream(snapshotFile.getSessionID(), fileName));
```

Comment:
Are you re-opening the snapshot file that you're currently writing to, so you can generate the hash? How can you guarantee everything has been written to it? If so, can you close the file first (i.e. after the try block) and read the hash separately?